### PR TITLE
feat(eslint-plugin): add `consistent-type-definitions` rule

### PR DESCRIPTION
**Pull request by otofu-square**
ref #142 

~~Adds `prefer-type-alias` rule which is the inverse of `prefer-interface`~~

Adds `consistent-type-definitions` rule to consist type definitions `interface` or `type`.

### CODE REVIEW

**Comment by bradzacher** [comment]
Hey @otofu-square - thanks for doing this!

Could you please "merge" the two rules (your new `prefer-type-alias`, and `prefer-interface`) together?
I suggest the new rule be named: `consistent-type-definitions`.

---

One goal of eslint is to prevent the ability to get yourself into inconsistent states via config.
If we create this as a separate rule, then a user could turn on both rules, and have one rule telling them to use `type`s instead of `interface`s, and the other telling them to use `interface`s instead of `type`s.
(as an aside - down the line we'll have to look at merging in `no-type-alias` as well, but don't worry about that for now!)

---

You won't be able to delete `prefer-interface`, but you can remove its row from the root `README.md`, and mark it as deprecated by adding the following properties to its meta:
```json
{
    "deprecated": true,
    "replacedBy": "consistent-type-definitions"
}
```

**Comment by otofu-square** [comment]
@bradzacher 
Thanks for the suggestion!
The naming `consistent-type-definitions` looks great.
I'll give it a try to merge two rules.

> If we create this as a separate rule, then a user could turn on both rules, and have one rule telling them to use types instead of interfaces, and the other telling them to use interfaces instead of types.

I see.
That’s very helpful for me, I learn something important about creating the eslint rules.

**Comment by otofu-square** [comment]
@bradzacher 
Do you think what would be a good option schema for `consistent-type-definitions` rule to switch the setting between `prefer-interface` and `prefer-type-alias`?
I came up with the schema like below, but I do not know if it is the best.

```
...
schema: [
  {
    type: 'object',
    properties: {
      preferDeclaration: {
        enum: [
          'type',
          'interface',
        ],
      },
    }
  }
]
```

**Comment by bradzacher** [comment]
In this case, I would say that generally the preference for rule config is to have the first argument be a string, followed by an object for additional config items.

I.e. As a tuple type it'd be
```ts
type Options = [
  ('type' | 'interface')?,
  {
    opt1 ?: value,
  }?,
];
```

File: packages/eslint-plugin/typings/ts-eslint.d.ts [hunk 1]
```diff
@@ -262,7 +262,7 @@ declare module 'ts-eslint' {
     /**
      * The name of the rule this rule was replaced by, if it was deprecated.
      */
-    replacedBy?: string;
+    replacedBy?: string[];
```
Comment (otofu-square) [inline packages/eslint-plugin/typings/ts-eslint.d.ts]: It seems that `replacedBy` has `string[]` type.
https://eslint.org/docs/developer-guide/working-with-rules#rule-basics

**Comment by otofu-square** [comment]
@bradzacher 
Please review my implementation once.
I'll give it a try to write docs if that is ok.

File: packages/eslint-plugin/src/rules/consistent-type-definisions.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import { TSESTree } from '@typescript-eslint/typescript-estree';
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/consistent-type-definisions.ts]: your filename is wrong - it should be `consistent-type-definitions`

File: packages/eslint-plugin/tests/rules/consistent-type-definitions.test.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    `interface A {}`,
+    `interface A extends B { x: number; }`,
+    `type U = string;`,
+    `type V = { x: number; } | { y: string; };`,
+    `
+type Record<T, U> = {
+    [K in T]: U;
+}
+`,
+  ],
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/consistent-type-definitions.test.ts]: you should also add tests which set the option to one or the other, to validate your options work as expected
i.e.
```ts
{
  code: `type A = { x: number }`,
  options: ['type'],
}
```

File: packages/eslint-plugin/src/rules/consistent-type-definisions.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+              ) {
+                fixes.push(fixer.remove(afterToken));
+              }
+
+              return fixes;
+            },
+          });
+        }
+      },
+      TSInterfaceDeclaration: node => {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/consistent-type-definisions.ts]: nitpick
```suggestion
      TSInterfaceDeclaration(node) {
```

**Comment by otofu-square** [comment]
> you should avoid force pushing when possible.
if you force push, then github cannot track your history, as your force push obliterates it!

It makes sense to me!

**Comment by otofu-square** [comment]
@bradzacher 
Have done writing the doc for this rule 👍

**Comment by leoyli** [comment]
Oh, I really love to have this rule come alive!!! :+1:

**Comment by bradzacher** [comment]
@otofu-square - there are two merge conflicts which need resolving (github thinks they're too "complex", so I can't do them for you via the github web UI).

Once they're resolved I will merge to master.

**Comment by otofu-square** [comment]
@bradzacher 
Should I leave the prefer-interface docs?
CI checks have failed because of that docs does not exist.

**Comment by bradzacher** [comment]
@otofu-square - that looks like a bug in the documentation checker! I'll look into it tomorrow morning.

**Pull request merged**